src/lib/shutdown/shutdown.cpp

```diff
@@ -113,26 +119,96 @@ static uint16_t shutdown_counter = 0; ///< count how many times the shutdown wor
 
 static uint8_t shutdown_args = 0;
 
-static constexpr int max_shutdown_hooks = 1;
-static shutdown_hook_t shutdown_hooks[max_shutdown_hooks] = {};
+static constexpr int max_shutdown_hooks = 10;
+static shutdown_hook_t shutdown_hooks[max_shutdown_hooks] = {0};
 
 static hrt_abstime shutdown_time_us = 0;
 static constexpr hrt_abstime shutdown_timeout_us =
 	5_s; ///< force shutdown after this time if modules do not respond in time
 
+class ShutdownAckCallback : public uORB::SubscriptionCallback
+{
+public:
+	ShutdownAckCallback() :
+		uORB::SubscriptionCallback(ORB_ID(shutdown_ack))
+	{}
+
+	void call()
+	{
+		if (updated()) {
+			shutdown_ack_s ack_msg;
+			copy(&ack_msg);
+
+			if (ack_msg.pid > 0) {
+				pthread_mutex_lock(&shutdown_mutex);
+
+				for (int i = 0; i < max_shutdown_hooks; ++i) {
+					if (shutdown_hooks[i] == ack_msg.pid) {
+						PX4_DEBUG("Shutdown ack from pid %i\n", ack_msg.pid);
+						shutdown_hooks[i] = -1;
+						pthread_mutex_unlock(&shutdown_mutex);
+						return;
+					}
+				}
+
+				pthread_mutex_unlock(&shutdown_mutex);
+
+			} else {
+				PX4_WARN("Shutdown ack from unknown pid: %i\n", ack_msg.pid);
+			}
+		}
+
+		return;
+	}
+};
+
+static ShutdownAckCallback *shutdown_ack_callback{nullptr};
+
+static uORB::Publication<shutdown_event_s> *shutdown_event{nullptr};
+
 int px4_register_shutdown_hook(shutdown_hook_t hook)
 {
+	// create and advertise shutdown event topic if not done yet
+	if (!shutdown_event) {
```

Comment:
This is a race condition? if two threads register a hook simultaneously, you may end up creating the publication twice.